src/charm.py

```diff
@@ -97,92 +100,116 @@ def _config(self) -> str:
             self.unit.status = WaitingStatus("Waiting to connect to Kratos container")
             return
 
-        if not self.model.relations["pg-database"]:
-            logger.error("Missing required relation with postgresql")
-            self.model.unit.status = BlockedStatus("Missing required relation with postgresql")
+        self.unit.status = MaintenanceStatus("Configuring/deploying resources")
+
+        with open("src/identity.default.schema.json", encoding="utf-8") as schema_file:
+            schema = schema_file.read()
+            self._container.push(self._identity_schema_file_path, schema, make_dirs=True)
+        self._container.push(self._config_file_path, self._config, make_dirs=True)
+        self._container.add_layer(self._container_name, self._pebble_layer, combine=True)
+        logger.info("Pebble plan updated with new configuration, replanning")
+        self._container.replan()
+
+        # in case container was terminated unexpectedly
+        if self.database.is_database_created():
+            self._container.start(self._container_name)
+            self.unit.status = ActiveStatus()
             return
 
-        if not self.database.is_database_created():
-            event.defer()
-            logger.info("Missing database details. Deferring event.")
+        if self.model.relations["pg-database"]:
             self.unit.status = WaitingStatus("Waiting for database creation")
+        else:
+            self.unit.status = BlockedStatus("Missing postgres database relation")
+
+    def _on_database_created(self, event) -> None:
+        """Event Handler for database created event."""
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Kratos container. Deferring event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Kratos container")
             return
 
+        self.unit.status = MaintenanceStatus(
+            "Configuring container and resources for database connection"
+        )
+
         try:
-            self._update_layer()
-        except ChangeError as err:
-            logger.error(str(err))
-            self.unit.status = BlockedStatus("Failed to replan")
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            event.defer()
+            self.unit.status = WaitingStatus("Waiting for Kratos service")
+            logger.info("Kratos service is absent. Deferring database created event.")
             return
 
-        if not self.unit.is_leader():
-            return
+        logger.info("Updating Kratos config and restarting service")
+        self._container.push(self._config_file_path, self._config, make_dirs=True)
 
-        self._run_sql_migration()
+        if self.unit.is_leader() and not self._run_sql_migration():
+            self.unit.status = BlockedStatus("Database migration failed.")
```

Comment:
Would you like to add an action for running the migration or handle it in another PR?